pkg/storage/iterators_test.go

```diff
@@ -104,3 +104,116 @@ func ExampleNewFilteredTupleKeyIterator() {
 	fmt.Println(filtered)
 	// Output: [document:doc1#editor@user:elbuo]
 }
+
+func TestConditionsFilteredTupleKeyIterator(t *testing.T) {
+	filter := func(tupleKey *openfgav1.TupleKey) (bool, error) {
+		switch tupleKey.GetCondition().GetName() {
+		case "condition1":
+			return true, nil
+		case "condition2":
+			return false, nil
+		default:
+			return false, fmt.Errorf("unknown condition: %s", tupleKey.GetCondition().GetName())
+		}
+	}
+
+	t.Run("no_error", func(t *testing.T) {
+		tuples := []*openfgav1.TupleKey{
+			tuple.NewTupleKeyWithCondition("document:doc1", "viewer", "user:jon", "condition1", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:elbuo", "condition2", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:maria", "condition1", nil),
+		}
+		iter := NewConditionsFilteredTupleKeyIterator(NewStaticTupleKeyIterator(tuples), filter)
+		var actual []*openfgav1.TupleKey
+		for {
+			tk, err := iter.Next(context.Background())
+			if err != nil {
+				if errors.Is(err, ErrIteratorDone) {
+					break
+				}
+				require.Fail(t, "no error was expected")
+			}
+
+			actual = append(actual, tk)
+		}
+		expected := []*openfgav1.TupleKey{tuples[0], tuples[2]}
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("has_some_valid_but_middle_invalid", func(t *testing.T) {
+		tuples := []*openfgav1.TupleKey{
+			tuple.NewTupleKeyWithCondition("document:doc1", "viewer", "user:jon", "condition1", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:elbuo", "condition3", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:maria", "condition1", nil),
+		}
+		iter := NewConditionsFilteredTupleKeyIterator(NewStaticTupleKeyIterator(tuples), filter)
+		var actual []*openfgav1.TupleKey
+		for {
+			tk, err := iter.Next(context.Background())
+			if err != nil {
+				if errors.Is(err, ErrIteratorDone) {
+					break
+				}
+				require.Fail(t, "no error was expected")
+			}
+
+			actual = append(actual, tk)
+		}
+		expected := []*openfgav1.TupleKey{tuples[0], tuples[2]}
+		require.Equal(t, expected, actual)
+	})
+
+	t.Run("has_some_valid_but_last_invalid", func(t *testing.T) {
+		tuples := []*openfgav1.TupleKey{
+			tuple.NewTupleKeyWithCondition("document:doc1", "viewer", "user:jon", "condition1", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:elbuo", "condition2", nil),
+			tuple.NewTupleKeyWithCondition("document:doc1", "editor", "user:maria", "condition3", nil),
+		}
+		iter := NewConditionsFilteredTupleKeyIterator(NewStaticTupleKeyIterator(tuples), filter)
+		var actual []*openfgav1.TupleKey
+		for {
+			tk, err := iter.Next(context.Background())
+			if err != nil {
+				if errors.Is(err, ErrIteratorDone) {
```

Comment:
onceValid -> if there are any valid, we should not emit error.

Reply:
but why? in the context of a Check request, if we encountered one valid tuple _that doesn't lead to a `allowed=true` outcome, and one that we couldn't evaluate, we should return the evaluation error no? because at this point in execution in the iterator, you have no idea how the consumer used that first valid tuple?

Reply:
updated comments to reflect when we return error.

Reply:
Never mind, I just understood - this is trying to mimic the behavior of the old way of doing things where we only return the error if we couldn't find any tuples